index.js

```diff
@@ -137,11 +138,58 @@ class Installer {
           this.pkgCount++
           return this
         })
+        .catch(e => {
+          if (child.optional) {
+            this.pkgCount++
+            child.failed = true
+            return rimraf(childPath)
+          } else {
+            throw e
+          }
+        })
       })
       return child.pending
     }, { concurrency: 50 })
   }
 
+  // A cute little mark-and-sweep collector!
+  garbageCollect (tree) {
+    const liveDeps = new Set()
+    const installer = this
+    mark(tree)
+    return sweep(tree)
+
+    function mark (tree) {
+      for (let dep of tree.dependencies.values()) {
+        if (dep.seenByMark) { continue }
+        dep.seenByMark = true
+        if (!dep.optional || !dep.failed) {
+          liveDeps.add(dep)
```

Comment:
You may be able to eliminate this mark phase if you add to a failed Set instead of doing dep.failed during the extract phase...